src/hotspot/share/gc/g1/g1ConcurrentMarkThread.cpp

```diff
@@ -309,14 +315,26 @@ void G1ConcurrentMarkThread::full_concurrent_cycle_do() {
   phase_clear_bitmap_for_next_mark();
 }
 
-void G1ConcurrentMarkThread::concurrent_cycle_end() {
+void G1ConcurrentMarkThread::concurrent_undo_cycle_do() {
+  HandleMark hm(Thread::current());
+  ResourceMark rm;
+
+  // We can (and should) abort if there has been a concurrent cycle abort for
+  // some reason.
+  if (_cm->has_aborted()) { return; }
```

Comment:
I think more doc could be added here to explain in what situation this condition becomes true. I don't see sth similar in `concurrent_mark_cycle_do`; why is it not needed there? I believe the answers could be placed in the comments to help future readers better understand the code.

Reply:
Currently the only reason for an abort is a full gc, calling `G1CollectedHeap::abort_concurrent_cycle`. I would like to avoid putting all callers of that method in this comment because they might change.

`concurrent_mark_cycle_do()` has one of these `if (_cm->has_aborted())` checks at the end of every phase implicitly, i.e. the many

`if (some_phase) return;`

encapsulate it. G1 can't abort before the root scan phase in `concurrent_mark_cycle_do`, because of this reason in the comment in `G1CollectedHeap::abort_concurrent_cycle`:

  // If we start the [full gc] compaction before the CM threads finish
  // scanning the root regions we might trip them over as we'll
  // be moving objects / updating references. So let's wait until
  // they are done. By telling them [root region scan] to abort, they should complete
  // early.

and we can't prevent the pause ("join an STS") in root region scanning or before because of the existing comment in `G1CollectedHeap::abort_concurrent_cycle`.

I can add a reference to the `G1CollectedHeap::abort_concurrent_cycle` comment to the one in `concurrent_mark_cycle_do` - would that satisfy your concern?

Reply:
> I can add a reference to the...

Yes, thank you very much.